### PR TITLE
Create package

### PR DESCRIPTION
**Pull request by antonelepfl**
Basically after running the simulation, we have some files and this package helps with the transfer of those files to a Jupyter Notebook with help of Unicore API.
We have mainly 2 Classes in this package:
- The one that generate a list of the jobs that we launched so we can select (JobFinder)
- The machinery to use unicore to fetch those results in an accessible place for the notebook (Results)

### CODE REVIEW

**Comment by antonelepfl** [comment]
Points to be considered:
- Should we hide the circuit directory on the package?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+UNICORE_ENDPOINT= 'https://bspsa.cineca.it/advanced/pizdaint/rest/core'
+
+
+class Results:
+
+    DEFAULT_CIRCUIT_DIR = '/home/data-bbp/20191017/'
+    DEFAULT_WD_BASE = '/home/simulation-results';
+    ORIGIN_CIRCUIT_PATH = '/store/hbp/ich002/antonel/O1/20191017/'
+    ORIGIN_DIR_PATH = '/scratch/snx3000/unicore/FILESPACE/'
+    job_base = join(UNICORE_ENDPOINT, 'jobs')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: This one is hard coded as well, so it should be upper case as well if I follow your own convention

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    files_list = None
+    circuit_dir = None
+    local_dir = None
+    blueconfig = None
+
+    def __init__(self, token, sim_id, base_working_directory=None, new_circuit_path=None):
+        if base_working_directory == None:
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        if base_working_directory is None:
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    DEFAULT_CIRCUIT_DIR = '/home/data-bbp/20191017/'
+    DEFAULT_WD_BASE = '/home/simulation-results';
+    ORIGIN_CIRCUIT_PATH = '/store/hbp/ich002/antonel/O1/20191017/'
+    ORIGIN_DIR_PATH = '/scratch/snx3000/unicore/FILESPACE/'
+    job_base = join(UNICORE_ENDPOINT, 'jobs')
+    files_list = None
+    circuit_dir = None
+    local_dir = None
+    blueconfig = None
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: They should be initialized in the `__init__`

**Review by wizmer — COMMENTED** [review]
I'm not done yet but can you run pylint on the directory and fix those ? Thanks

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import pandas as pd
+
+UNICORE_ENDPOINT= 'https://bspsa.cineca.it/advanced/pizdaint/rest/core'
+
+
+class Results:
+
+    DEFAULT_CIRCUIT_DIR = '/home/data-bbp/20191017/'
+    DEFAULT_WD_BASE = '/home/simulation-results';
+    ORIGIN_CIRCUIT_PATH = '/store/hbp/ich002/antonel/O1/20191017/'
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: Is it okay that it points to your home dir ?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    job_base = join(UNICORE_ENDPOINT, 'jobs')
+    files_list = None
+    circuit_dir = None
+    local_dir = None
+    blueconfig = None
+
+    def __init__(self, token, sim_id, base_working_directory=None, new_circuit_path=None):
+        if base_working_directory == None:
+            base_working_directory = self.DEFAULT_WD_BASE
+        if new_circuit_path == None:
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        if new_circuit_path is None:
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            print('[Error] Simulation ID not provided')
+            return
+        self.retrieve_sim_info(token, sim_id)
+        self.get_sim_results()
+
+    def retrieve_sim_info(self, token, sim_id=None):
+        tr = unicore_client.Transport(token)
+        job_url = join(self.job_base, sim_id)
+        job = unicore_client.Job(tr, job_url)
+        print('Fetching results of [{}]. Please wait ...'.format(job.properties['name']))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: Use a logger

```python
import logging

L = logging.getLogger('dynamic-analysis-pkg')

L.info('Fetching results of [{}]. Please wait ...'.format(job.properties['name']))
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+        if not isdir(new_circuit_path):
+            print('[ERROR] Circuit path does not exists')
+        else:
+            self.circuit_dir = new_circuit_path
+
+        self.ORIGIN_DIR_PATH = join(self.ORIGIN_DIR_PATH, sim_id)
+
+    def download_file_to_storage(self, file_name):
+        # function to download the file and add it to kernel file system
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        """function to download the file and add it to kernel file system"""
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        self.local_dir = join(working_directory, sim_id)
+        if not exists(self.local_dir):
+            makedirs(self.local_dir)
+
+        if not isdir(new_circuit_path):
+            print('[ERROR] Circuit path does not exists')
+        else:
+            self.circuit_dir = new_circuit_path
+
+        self.ORIGIN_DIR_PATH = join(self.ORIGIN_DIR_PATH, sim_id)
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: I don't think this is what you intended to do.

```python
class A:
    VALUE_COMMON_TO_ALL_INSTANCES = None

    def __init__(self):
       # Things specific to this instance (the most general case) have to be defined in the `__init__(self)`
        self.value_specific_to_this_instance_of_the_class = None
      
      # This will change the values for all instances of the class
       VALUE_COMMON_TO_ALL_INSTANCES = 2
```

TLDR: use another variable
```suggestion
        self.dir_path = join(self.ORIGIN_DIR_PATH, sim_id)
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+    def download_file_to_storage(self, file_name):
+        # function to download the file and add it to kernel file system
+        file_output_path = join(self.local_dir, file_name)
+        if exists(file_output_path):
+            print('- [{}] file already exists. Skipping download.'.format(file_name))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
            L.info('- [{}] file already exists. Skipping download.'.format(file_name))
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+            return
+
+        print('- Fetching [{}] ...'.format(file_name))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        L.info('- Fetching [{}] ...'.format(file_name))
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            with open(file_output_path, 'w') as fd:
+                fd.write(writable_content)
+
+        else:
+            x.download(str(file_name)) # moves to home always
+            move(file_name, file_output_path)
+
+        print('- [{}] downloaded'.format(file_name))
+
+    def download_blueconfig(self, file_name='BlueConfig'):
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: This function is just the other one but with a default argument. I'm not sure it is needed.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            file_content = x.raw().read()
+            bc_str = file_content.decode('utf-8')
+            writable_content = bc_str.replace(self.ORIGIN_CIRCUIT_PATH, self.circuit_dir)
+            writable_content = writable_content.replace(self.ORIGIN_DIR_PATH, self.local_dir)
+
+            with open(file_output_path, 'w') as fd:
+                fd.write(writable_content)
+
+        else:
+            x.download(str(file_name)) # moves to home always
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: What is the `type` of `file_name` ? Why do you need to cast it to `str` ?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            move(file_name, file_output_path)
+
+        print('- [{}] downloaded'.format(file_name))
+
+    def download_blueconfig(self, file_name='BlueConfig'):
+        self.download_file_to_storage(file_name)
+
+    def download_report(self):
+        reports = [x for x in self.files_list if '.bbp' in x]
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: If you are just looking for the end of the name then:
```suggestion
        reports = [x for x in self.files_list if x.endswith('.bbp')]
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            writable_content = writable_content.replace(self.ORIGIN_DIR_PATH, self.local_dir)
+
+            with open(file_output_path, 'w') as fd:
+                fd.write(writable_content)
+
+        else:
+            x.download(str(file_name)) # moves to home always
+            move(file_name, file_output_path)
+
+        print('- [{}] downloaded'.format(file_name))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        info('- [{}] downloaded'.format(file_name))
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        self.files_list = storage.listdir()
+
+    def get_sim_results(self):
+        self.download_blueconfig()
+        self.download_report()
+        self.download_out_dat()
+        print('Result were saved at: {}'.format(self.local_dir))
+
+    def define_paths(self, sim_id, working_directory, new_circuit_path):
+        self.local_dir = join(working_directory, sim_id)
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: In python all attributes should be initialized in the `__init__(self)`.
Since this function is only called in `__init__` I would simply remove it and put its content in the `__init__` body.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        if not exists(self.local_dir):
+            makedirs(self.local_dir)
+
+        if not isdir(new_circuit_path):
+            print('[ERROR] Circuit path does not exists')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
            L.error('[ERROR] Circuit path does not exists')
```
Shouldn't you raise here ?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
+
+from os import makedirs
+from os.path import join, exists, isdir, basename
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: You should take the habit of using `os.path` big brother built in in python 3:
```suggestion
from pathlib import Path
```
Then wrap all you path into `Path` obects:
```python
my_path = Path(my_path_which_was_a_string)

# replace 
# exists by:
my_path.exists()

# mkdirs by:
my_path.mkdir() # there is a flag if you want/don't want to raise when it already exists

# join by:
my_sub_dir_path = my_path / 'sub-dir'

# is_dir by:
my_path.is_dir()

# basename by:
my_path.name
```

This is probably the best addition to python3, read more at:
https://docs.python.org/3/library/pathlib.html

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+
+        print('- [{}] downloaded'.format(file_name))
+
+    def download_blueconfig(self, file_name='BlueConfig'):
+        self.download_file_to_storage(file_name)
+
+    def download_report(self):
+        reports = [x for x in self.files_list if '.bbp' in x]
+        if not reports:
+            print('No reports were found')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
            L.info('No reports were found')
```
or maybe `L.warn`

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+class FetchMultipleResults:
+
+    values = []
+
+    def __init__(self, token, sim_list_ids, base_working_directory=None):
+
+        total_sims = len(sim_list_ids)
+        for idx, sim in enumerate(sim_list_ids):
+            print('({}/{})'.format(idx + 1, total_sims))
+            self.values.append(Results(token, sim, base_working_directory))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: You can have a progress bar with:
```python
from tqdm import tqdm
```

```suggestion
    for sim in tqdm(sim_list_ids):
            self.values.append(Results(token, sim, base_working_directory))
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+
+
+class JobFinder:
+
+    jobs = []
+    sorted_jobs = None
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: Should be in `__init__`

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+
+    def __init__(self, token):
+        tr = unicore_client.Transport(token)
+        client = unicore_client.Client(tr, UNICORE_ENDPOINT)
+        self.jobs = client.get_jobs()
+
+    def show_list(self, amount_jobs_to_show=None):
+        full_jobs_list = []
+        if amount_jobs_to_show is None:
+            amount_jobs_to_show = 1000
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
    def show_list(self, amount_jobs_to_show=1000):
        full_jobs_list = []
```
This gives the same result except if you are expecting `amount_jobs_to_show = None` to be passed from the outside.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+            raw_time = job.properties['submissionTime']
+            full_time = datetime.strptime(raw_time, '%Y-%m-%dT%H:%M:%S%z')
+            date = full_time.strftime('%d-%m-%Y %H:%M')
+            name = job.properties['name']
+            job_id = basename(job.properties['_links']['self']['href'])
+            full_jobs_list.append([index, name, date, job_id])
+            print('Fetching jobs ({}/{})'.format(index + 1, amount_jobs_to_show or len(jobs) + 1), end = '\r')
+
+        columns = ['index', 'name', 'date', 'job_id']
+        self.sorted_results = pd.DataFrame(full_jobs_list, columns=columns).sort_values(by=['date'])
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        columns = ['name', 'date', 'job_id']
        self.sorted_results = pd.DataFrame(full_jobs_list, columns=columns).reset_index().sort_values(by=['date'])
```
But check that it works, I'm not so sure :D

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+            amount_jobs_to_show = 1000
+
+        for index, job in enumerate(self.jobs[:amount_jobs_to_show]):
+            raw_time = job.properties['submissionTime']
+            full_time = datetime.strptime(raw_time, '%Y-%m-%dT%H:%M:%S%z')
+            date = full_time.strftime('%d-%m-%Y %H:%M')
+            name = job.properties['name']
+            job_id = basename(job.properties['_links']['self']['href'])
+            full_jobs_list.append([index, name, date, job_id])
+            print('Fetching jobs ({}/{})'.format(index + 1, amount_jobs_to_show or len(jobs) + 1), end = '\r')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        for job in tqdm(self.jobs[:amount_jobs_to_show]):
            raw_time = job.properties['submissionTime']
            full_time = datetime.strptime(raw_time, '%Y-%m-%dT%H:%M:%S%z')
            full_jobs_list.append([job.properties['name'],
                                               full_time.strftime('%d-%m-%Y %H:%M'),
                                               Path(job.properties['_links']['self']['href']).name])
```
See next comment for why you no longer need to keep the index.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+
+        columns = ['index', 'name', 'date', 'job_id']
+        self.sorted_results = pd.DataFrame(full_jobs_list, columns=columns).sort_values(by=['date'])
+
+        pd.options.display.max_columns = None
+        pd.options.display.width=None
+
+        print('\n') # to stop the \r
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: Wut ??

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+        self.sorted_results['index'] = range(0, amount_jobs_to_show)
+        print(tabulate(self.sorted_results[showing_columns], headers=showing_columns, showindex=False))
+
+
+    def find_id_by_index(self):
+        job_selection_index = input('Please select the index that you want to import:')
+        try:
+            selected_index = int(job_selection_index)
+        except:
+            print('ERROR: enter correct index')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: You want to raise here, no ?

```suggestion
            raise ValueError('ERROR: enter correct index')
```
?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+
+        selection = self.sorted_results.iloc[int(selected_index)]
+        job_id = selection['job_id']
+        print('ID: {}'.format(job_id))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: Starting at python 3.6 you can use the new string formatting:
```suggestion
        L.info(f'ID: {job_id}')
```

File: dynamic_analysis_pkg/setup.py [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+      version='0.1.7',
+      description='Machinary to fetch simulation results to a Jupyter Notebook',
+      author='Stefano Antonel',
+      author_email='stefano.antonel@epfl.ch',
+      license='MIT',
+      packages=['dynamic_analysis_pkg'],
+      install_requires=[
+        'pyunicore',
+        'tabulate',
+      ],
```
Comment (wizmer) [inline dynamic_analysis_pkg/setup.py:14]:
```suggestion
        'tqdm>=X.X.X',
      ],
```
And add some minimum version number for each dependencies:

File: dynamic_analysis_pkg/setup.py [hunk 28]
```diff
@@ -0,0 +1,14 @@
+from setuptools import setup
+
+setup(name='dynamic_analysis_pkg',
```
Comment (wizmer) [inline dynamic_analysis_pkg/setup.py]:
```suggestion
setup(name='dynamic-analysis-pkg',
```

File: dynamic_analysis_pkg/setup.py [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+      version='0.1.7',
+      description='Machinary to fetch simulation results to a Jupyter Notebook',
+      author='Stefano Antonel',
+      author_email='stefano.antonel@epfl.ch',
```
Comment (wizmer) [inline dynamic_analysis_pkg/setup.py]: Up to you but I usually put:
```suggestion
    author="BlueBrain NSE",
    author_email="bbp-ou-nse@groupes.epfl.ch",
```
But it might be different for the mooc. I don't know.

Comment (asanin-epfl) [reply dynamic_analysis_pkg/setup.py]: Should we apply here our Python guidelines?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+            self.circuit_dir = new_circuit_path
+
+        self.local_dir = Path(working_directory) / Path(sim_id)
+        self.local_dir.mkdir(parents=True, exist_ok=True)
+
+        self.hpc_dir_path = Path(self.HPC_DIR_BASE) / Path(sim_id)
+
+    def download_file_to_storage(self, file_name):
+        ''' function to download the file and add it to kernel file system '''
+        file_output_path = Path(self.local_dir) / Path(file_name)
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: The only place where you should need the `Path` constructor is when you define the paths.
On this line for examples, `self.local_dir` is already a `Path` object so you can simply:
```suggestion
        file_output_path = self.local_dir / file_name
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 31]
```diff
@@ -0,0 +1,167 @@
...
+        self.download_file_to_storage('out.dat')
+        L.info('Result were saved at: %s', self.local_dir)
+
+    def define_paths(self, sim_id, working_directory, new_circuit_path):
+        if not Path(new_circuit_path).is_dir():
+            raise ValueError('[ERROR] Circuit path does not exists')
+        else:
+            self.circuit_dir = new_circuit_path
+
+        self.local_dir = Path(working_directory) / Path(sim_id)
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: You don't need what is after the slash to be a `Path`, it can be a string.
```suggestion
        self.local_dir = Path(working_directory) / sim_id
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 32]
```diff
@@ -0,0 +1,167 @@
...
+
+class AResults:
+    '''
+    Based on a simulation id, fetch the results using Unicore API
+    '''
+    DEFAULT_CIRCUIT_DIR = '/home/data-bbp/20191017/'
+    DEFAULT_WD_BASE = '/home/simulation-results'
+    ORIGIN_CIRCUIT_PATH = '/store/hbp/ich002/antonel/O1/20191017/'
+    HPC_DIR_BASE = '/scratch/snx3000/unicore/FILESPACE/'
+    JOB_BASE = os.path.join(UNICORE_ENDPOINT, 'jobs')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: Define all those as `Path`. It will make most of the following `Path` calls unnecessary.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 33]
```diff
@@ -0,0 +1,167 @@
...
+        if new_circuit_path is None:
+            new_circuit_path = self.DEFAULT_CIRCUIT_DIR
+
+        self.define_paths(sim_id, base_working_directory, new_circuit_path)
+        self.fetch_results(token, sim_id)
+        self.blueconfig = Path(self.local_dir) / 'BlueConfig'
+
+    def fetch_results(self, token, sim_id=None):
+        if not sim_id:
+            raise ValueError('[Error] Simulation ID not provided')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: This is really weird. You allow for a default parameter `sim_id=None`, but if passed, you raise :D
What's the point of the default then ?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 34]
```diff
@@ -0,0 +1,167 @@
...
+
+        self.retrieve_sim_info(token, sim_id)
+        self.get_sim_results()
+
+    def retrieve_sim_info(self, token, sim_id=None):
+        tr = unicore_client.Transport(token)
+        job_url = os.path.join(self.JOB_BASE, sim_id)
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py:56]: Since in your next patchset, `self.JOB_BASE` will be a `Path`:
```suggestion
        job_url = self.JOB_BASE / sim_id
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 35]
```diff
@@ -0,0 +1,167 @@
...
+    def __init__(self, token):
+        self.sorted_jobs = None
+        self.sorted_results = None
+        tr = unicore_client.Transport(token)
+        client = unicore_client.Client(tr, UNICORE_ENDPOINT)
+        self.jobs = client.get_jobs()
+
+    def show_list(self, amount_jobs_to_show=1000):
+        full_jobs_list = []
+        fetched_index = 0
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: You still don't need this.
You can simply use `pandas.DataFrame.reset_index()` to return you a dataframe where the index has been converted as a column.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py (hunk 34 above)
Comment (antonelepfl) [reply dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py:56]: For this one I have a question..
If I use Path with and URL (in this case UNICORE_ENDPOINT) it removes `//` and set `/` and that will mess up the API call, that is why I was using the `os.path.join` do you know any other way?

Comment (wizmer) [reply dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py:56]: I see. Yes, I think you are right. I don't think `Path` is intended to be used with URLs.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 36]
```diff
@@ -0,0 +1,157 @@
...
+        file_output_path = self.local_dir / file_name
+        if file_output_path.exists():
+            L.info('- [%s] file already exists. Skipping download.', file_name)
+            return
+
+        L.info('- Fetching [%s] ...', file_name)
+        try:
+            current_file = self.files_list[file_name]
+        except:
+            raise KeyError('The file [{}] is not present on the simulation results'.format(file_name))
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        if file_name not in self.files_list:
            raise KeyError('The file [{}] is not present on the simulation results'.format(file_name))
        current_file = self.files_list[file_name]
```

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 37]
```diff
@@ -0,0 +1,157 @@
...
+            current_file.download(file_name) # moves to home always
+            move(file_name, file_output_path)
+
+        L.info('- [%s] downloaded', file_name)
+
+    def download_report(self):
+        reports = [x for x in self.files_list if x.endswith('.bbp')]
+        if not reports:
+            L.info('No reports were found')
+            return
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]: You don't need the `return`
The for loop will be skipped as `reports` is empty.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 38]
```diff
@@ -0,0 +1,157 @@
...
+            job_id = PurePath(job.properties['_links']['self']['href']).name
+            full_jobs_list.append([name, date, job_id, full_time, status])
+
+        columns = ['name', 'date', 'job_id', 'timestamp', 'status']
+        self.sorted_results = pd.DataFrame(full_jobs_list, columns=columns).sort_values(
+            by=["timestamp"]
+        ).reset_index()
+
+        showing_columns = ['name', 'date', 'status']
+        print(tabulate(self.sorted_results[showing_columns],
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py:142]: There is a `print` here. Did you keep it by design ?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 39]
```diff
@@ -0,0 +1,157 @@
...
+                       headers=showing_columns))
+
+    def find_id_by_index(self):
+        job_selection_index = input('Please select the index that you want to import:')
+        try:
+            selected_index = int(job_selection_index)
+        except:
+            raise ValueError('ERROR: enter correct index')
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py:150]: Should not you do a while loop where you requery for the correct input instead of raising ?

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py [hunk 40]
```diff
@@ -0,0 +1,157 @@
...
+
+        selection = self.sorted_results.iloc[int(selected_index)]
```
Comment (wizmer) [inline dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py]:
```suggestion
        selection = self.sorted_results.iloc[selected_index]
```

**Review by wizmer — APPROVED** [review]
Some minor comments, otherwise I'm happy.

File: dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py (hunk 38 above)
Comment (antonelepfl) [reply dynamic_analysis_pkg/dynamic_analysis_pkg/__init__.py:142]: Is the way to show the tabulate object
